### PR TITLE
Added files to support azure datalake storage

### LINKED ISSUE #292: Support for azure Data Lake Gen2 type storage
## 🚀 Feature Request
The current support for Azure only covers the blob type storage and not the Data Lake Gen2 type storage which is more generally used for AI workloads. These support hierarchical namespaces in comparison to the flat namespaces in the case of blob storage. Also these type of storage are optimized for Big Data analytics and training on large sets of data.

## Motivation

With PR #256, we can use Streaming with azure blob type storage. I tried it out and it worked seamlessly. But when I tried it with Data Lake type storage it crashed, mainly due to differences in the way these two are handled by the sdk. 

## Implementation
 The code additions to allow their usage is minimal:

```
   # Create a new session per thread
    service = DataLakeServiceClient(
        account_url=f"https://{os.environ['AZURE_ACCOUNT_NAME']}.dfs.core.windows.net",
        credential= os.environ['AZURE_ACCOUNT_ACCESS_KEY'])
    try:
        file_client = service.get_file_client(file_system=obj.netloc, file_path=obj.path.lstrip('/'))
        with open(local, 'wb') as my_blob:
            file_data = file_client.download_file()
            file_data.readinto(my_blob)
    except ResourceNotFoundError:
        raise FileNotFoundError(f'Object {remote} not found.')
    except Exception:
        raise
```

We need only add a new service client for Data Lakes and change the corresponding  _get_file_client_ method. Also  a corresponding uploader needs to be created.

The only reason I haven't created a PR yet is that it is not yet clear to me what may be the best way to differentiate between calling either of the clients. Can we have a flag for the storage type ( storage_type = blob  or storage_type = datalake), hat can be set by the user? Which other files may need changing because of this? Please suggest and I can contribute.

### PR DESCRIPTION
**Pull request by shivshandilya**
## Description of changes:
This PR adds functionality to support Azure DataLake type storage.

## Issue :
This is in response to the discussion in the issue #292.
Fixes #292.

### Upload
```
import numpy as np
from PIL import Image
from streaming import MDSWriter

# Local or remote directory in which to store the compressed output files
data_dir = "azure-dl://test-streaming"

# A dictionary mapping input fields to their data types
columns = {
    'image': 'jpeg',
    'class': 'int'
}

# Shard compression, if any
compression = 'zstd'

# Save the samples as shards using MDSWriter
with MDSWriter(out=data_dir, columns=columns, compression=compression) as out:
    for i in range(10000):
        sample = {
            'image': Image.fromarray(np.random.randint(0, 256, (32, 32, 3), np.uint8)),
            'class': np.random.randint(10),
        }
        out.write(sample)
```
image

### Download

```
from torch.utils.data import DataLoader
from streaming import StreamingDataset

# Remote path where full dataset is persistently stored
remote = "azure-dl://test-streaming"

# Local working dir where dataset is cached during operation
local = '/tmp'

# Create streaming dataset
dataset = StreamingDataset(local=local, remote=remote, shuffle=True)

# Let's see what is in sample #1337...
sample = dataset[1337]
img = sample['image']
cl = sample['class']

print(sample)
print(img, cl)
```

#### Output
```
(venv) azureuser@shivshandilya $ python test_streaming_download.py 

{'class': 6, 'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=32x32 at 0x7FA522C14910>}
<PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=32x32 at 0x7FA522C14910> 6
```

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [x] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#running-tests))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Comment by karan6181** [comment]
Hi @shivshandilya, thanks for creating the PR. With respect to your linting failures, are you running `pre-commit run --all-files` to ensure it complies with the linting rules ? Ref [link](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites).

**Comment by shivshandilya** [comment]
Hey @karan6181 , running the command `pre-commit run --all-files`, doesn't show any errors with the files I added but with some other older files like _docs/source/examples_ and all the _yaml_ files.

image

and 

image


</n></n></n>

There seems to be some eol erros that the pre-commit hooks detect. I am running this on a wsl environment with Ubuntu.
Any idea why this happens? maybe due to the difference in eol sequence in windows and linux(LF and CLRF)?

**Comment by shivshandilya** [comment]
Also strangely the pre-commit checks pass here, and not on my local machine😅.

**Comment by shivshandilya** [comment]
I am also running it on a wsl environment with Ubuntu, but I still have the error🤔.

File: streaming/base/storage/upload.py [hunk 1]
```diff
@@ -475,6 +477,81 @@ def check_bucket_exists(self, remote: str):
                 f'or check the bucket permission.',)
 
 
+class AzureDLUploader(CloudUploader):
```
Comment (knighton) [inline streaming/base/storage/upload.py:480]: Let's spell out "dl" (`AzureDataLakeUploader`, etc)

File: streaming/base/storage/download.py [hunk 2]
```diff
@@ -248,6 +248,38 @@ def download_from_azure(remote: str, local: str) -> None:
         raise
 
 
+def download_from_azuredl(remote: str, local: str) -> None:
+    """Download a file from remote Microsoft Azure to local.
```
Comment (karan6181) [inline streaming/base/storage/download.py:252]: nit: `Microsoft Azure` -> `Microsoft Azure Data Lake`

**Comment by karan6181** [comment]
Can you also update the description with the output showing that `upload` and `download` functionality works as expected ? Thanks!

**Pull request merged**